pgp/keysource.go

```diff
@@ -350,15 +562,73 @@ func (key *MasterKey) passphrasePrompt() func(keys []openpgp.Key, symmetric bool
 			k.PrivateKey.Decrypt([]byte(pass))
 			return []byte(pass), nil
 		}
-		return nil, fmt.Errorf("No key to unlock")
+
+		return nil, fmt.Errorf("no key to unlock")
 	}
 }
 
-// ToMap converts the MasterKey into a map for serialization purposes
-func (key MasterKey) ToMap() map[string]interface{} {
-	out := make(map[string]interface{})
-	out["fp"] = key.Fingerprint
-	out["created_at"] = key.CreationDate.UTC().Format(time.RFC3339)
-	out["enc"] = key.EncryptedKey
-	return out
+// loadRing attempts to load the keyring from the provided path.
+// Unsupported keys are ignored as long as at least a single valid key is
+// found.
+func loadRing(path string) (openpgp.EntityList, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	keyring, err := openpgp.ReadKeyRing(f)
+	if err != nil {
+		return nil, err
+	}
+	return keyring, nil
+}
+
+// fingerprintIndex indexes the openpgp.Entity objects from the given ring
+// by their fingerprint, and returns the result.
+func fingerprintIndex(ring openpgp.EntityList) map[string]openpgp.Entity {
+	fps := make(map[string]openpgp.Entity)
+	for _, entity := range ring {
+		fp := strings.ToUpper(hex.EncodeToString(entity.PrimaryKey.Fingerprint[:]))
+		if entity != nil {
+			fps[fp] = *entity
+		}
+	}
+	return fps
+}
+
+// gpgExec runs the provided args with the gpgBinary, while restricting it to
+// gnuPGHome. Stdout and stderr can be read from the returned buffers.
+// When the command fails, an error is returned.
+func gpgExec(gnuPGHome string, args []string, stdin io.Reader) (err error, stdout bytes.Buffer, stderr bytes.Buffer) {
+	if gnuPGHome != "" {
+		args = append([]string{"--no-default-keyring", "--homedir", gnuPGHome}, args...)
+	}
+
+	cmd := exec.Command(gpgBinary(), args...)
+	cmd.Stdin = stdin
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	err = cmd.Run()
+	return
+}
+
+// gpgBinary returns the GnuPG binary which must be used.
+// It allows for runtime modifications by setting the environment variable
+// SopsGpgExecEnv to the absolute path of the replacement binary.
+func gpgBinary() string {
+	binary := "gpg"
+	if envBinary := os.Getenv(SopsGpgExecEnv); envBinary != "" && filepath.IsAbs(envBinary) {
```

Comment:
@ajvb did a last comparison round, and we introduced this stricter behavior on the environment variable to ensure nothing magical is injected. Any concerns this may break things for people?